src/summarycode/file_cat.py

```diff
@@ -0,0 +1,1265 @@
+"""
+
+# Tracing flag
+TRACE = False
+
+
+def logger_debug(*args):
+    pass
+
+
+if TRACE:
+    import logging
+    import click
+
+    class ClickHandler(logging.Handler):
+        _use_stderr = True
+
+        def emit(self, record):
+            try:
+                msg = self.format(record)
+                click.echo(msg, err=self._use_stderr)
+            except Exception:
+                self.handleError(record)
+
+    logger = logging.getLogger(__name__)
+    logger.handlers = [ClickHandler()]
+    logger.propagate = False
+    logger.setLevel(logging.DEBUG)
+
+    def logger_debug(*args):
+        return logger.debug(" ".join(isinstance(a, str) and a or repr(a) for a in args))
+
+
+@post_scan_impl
+class FileCategorizer(PostScanPlugin):
+    """
+    Categorize a file.
+    """
+
+    resource_attributes = dict(
+        [
+            ("analysis_priority", String(help="Analysis priority.")),
+            ("file_category", String(help="File category.")),
+            ("file_subcategory", String(help="File subcategory.")),
+        ]
+    )
+
+    sort_order = 50
+
+    options = [
+        PluggableCommandLineOption(
+            ("--file-cat",),
+            is_flag=True,
+            default=False,
+            help="Categorize files.",
+            help_group=POST_SCAN_GROUP,
+            sort_order=50,
+        )
+    ]
+
+    def is_enabled(self, file_cat, **kwargs):
+        return file_cat
+
+    def process_codebase(self, codebase, file_cat, **kwargs):
+        if not file_cat:
+            return
+
+        for resource in codebase.walk(topdown=True):
+            category = categorize_resource(resource)
+            if not category:
+                continue
+            resource.analysis_priority = category.analysis_priority
+            resource.file_category = category.file_category
+            resource.file_subcategory = category.file_subcategory
+            resource.save(codebase)
+
+
+class Categorizer:
+    order = 0
+    analysis_priority = None
+    file_category = None
+    file_subcategory = None
+    category_notes = None
+    rule_applied = None
+
+    @classmethod
+    def categorize(cls, resource):
+        """
+        Return True if this Categorizer applies to this resource or False or None
+        otherwise.
+        """
+        raise NotImplementedError
+
+
+class ArchiveAndroid(Categorizer):
+    order = 10
+    analysis_priority = "1"
+    file_category = "archive"
+    file_subcategory = "Android"
+    rule_applied = "ArchiveAndroid"
+
+    @classmethod
+    def categorize(cls, resource):
+        return extension_in(resource.extension, [".apk", ".aar"])
+
+
+class ArchiveDebian(Categorizer):
+    order = 10
+    analysis_priority = "1"
+    file_category = "archive"
+    file_subcategory = "debian"
+    category_notes = "special type of archive"
+    rule_applied = "ArchiveDebian"
+
+    @classmethod
+    def categorize(cls, resource):
+        return extension_in(resource.extension, [".deb"])
```

Comment:
Since "extension in" is a common check made in several classes we could simplify this by creating a class variable for "extensions" and have a default "categorize" for an ExtensionCategorizer subclass of Categorizer?